src/backend/distributed/utils/metadata_cache.c

```diff
@@ -964,13 +973,246 @@ CitusHasBeenLoaded(void)
 			 * present during early stages of upgrade operation.
 			 */
 			DistPartitionRelationId();
+
+			/*
+			 * We also set installedExtensionVersion to NULL so that it will be re-read
+			 * in case of extension update.
+			 */
+			installedExtensionVersion = NULL;
 		}
 	}
 
+	if (extensionLoaded)
+	{
+		ErrorIfNewMajorVersionAvailable();
+		ErrorIfExtensionUpdateNeeded();
+	}
+
 	return extensionLoaded;
 }
 
 
+/*
+ * ErrorIfNewMajorVersionAvailable compares CITUS_EXTENSIONVERSION and currently
+ * available version from citus.control file. If they are not same in major or
+ * minor version numbers, this function errors out. It ignores the schema version.
+ */
+void
+ErrorIfNewMajorVersionAvailable(void)
+{
+	char *availableVersion = AvailableMajorVersion();
+
+	if (!CompareVersions(availableVersion, CITUS_EXTENSIONVERSION))
+	{
+		ereport(ERROR, (errmsg("server restart is needed because, loaded Citus binaries "
+							   "does not match the available extension version")));
+	}
+}
+
+
+/*
+ * ErrorIfExtensionUpdateNeeded compares CITUS_EXTENSIONVERSION and currently and
+ * catalog version from pg_extemsion catalog table. If they are not same in major or
+ * minor version numbers, this function errors out. It ignores the schema version.
+ */
+static void
+ErrorIfExtensionUpdateNeeded(void)
+{
+	char *installedVersion = InstalledExtensionVersion();
+
+	if (!CompareVersions(installedVersion, CITUS_EXTENSIONVERSION))
+	{
+		ereport(ERROR, (errmsg("\"ALTER EXTENSION citus UPDATE;\" is needed, because "
+							   "loaded Citus binaries does not match the installed "
+							   "extension version")));
+	}
+}
+
+
+/*
+ * CompareVersions compares given two versions. If they are same in major and
+ * minor version numbers, this function returns true. It ignores the schema version.
+ */
+bool
+CompareVersions(char *leftVersion, char *rightVersion)
```

Comment:
This function has a bug. Depending upon the order of arguments, it would return `true` for `6.1-` and `6.10-`.

You need to find the hyphen offset in _both_ strings and compare up to the farther point.